remyxai/utils/validators.py

```diff
@@ -0,0 +1,144 @@
+import os
+import re
+import logging
+import requests
+from typing import List, Tuple, Optional
+from huggingface_hub import HfFolder
+from remyxai.api.models import fetch_available_architectures
+
+def get_hf_token() -> Optional[str]:
+    """
+    Fetches the Hugging Face token from the user's environment.
+    Tries environment variable 'HF_TOKEN' first, then the Hugging Face cache.
+    """
+    hf_token = os.getenv('HF_TOKEN')
+    if hf_token:
+        return hf_token
+    hf_token = HfFolder.get_token()
+    return hf_token
+
+def get_headers(hf_token: Optional[str]):
+    headers = {}
+    if hf_token:
+        headers["Authorization"] = f"Bearer {hf_token}"
+    return headers
+
+def validate_model_architecture(
+    model_id: str, supported_archs: List[str], hf_token: Optional[str]
+) -> Tuple[bool, str]:
+    """
+    Validates if a model's architecture matches any known architectures from the server.
+    """
+    try:
+        api_url = f"https://huggingface.co/{model_id}/raw/main/config.json"
+        response = requests.get(api_url, headers=get_headers(hf_token), timeout=10)
+        response.raise_for_status()
+
+        config = response.json()
+        architectures = config.get("architectures", [])
+
+        if not supported_archs:
+            return False, "Supported architectures list is empty."
+
+        for architecture in architectures:
+            if architecture in supported_archs:
+                return True, f"Model '{model_id}' matches architecture: {architecture}"
+
+        return False, (
+            f"Model '{model_id}' does not match any supported architectures: {', '.join(supported_archs)}"
+        )
+
+    except requests.exceptions.HTTPError as e:
+        if e.response.status_code == 404:
+            return False, f"Model '{model_id}' does not exist or is not accessible."
+        else:
+            return False, f"HTTP error fetching model config for '{model_id}': {e}"
+    except requests.exceptions.RequestException as e:
+        return False, f"Request error fetching model config for '{model_id}': {e}"
+    except KeyError:
+        return False, f"Invalid configuration format for '{model_id}'"
+    except Exception as e:
+        logging.error(f"Unexpected error during architecture validation: {e}")
+        return False, f"Unexpected error during validation: {e}"
+
+def validate_model_size(model_id: str, max_size_billion: int = 8) -> Tuple[bool, str]:
+    """
+    Validates a model's size based on its repository name convention.
+    """
+    pattern = r"(\d+(\.\d+)?)([BM])"
+    match = re.search(pattern, model_id)
+
+    if not match:
+        return True, (
+            f"Model size for '{model_id}' could not be determined from the name; "
+            "assumed valid."
+        )
+
+    size_str, _, unit = match.groups()
+    size = float(size_str)
+
+    if unit.upper() == "M":
```

Comment:
parsing this way will fail for so many model ids, we need something more robust:
https://stackoverflow.com/questions/68086929/how-to-get-the-size-of-a-hugging-face-pretrained-model